File: python/Windows/shared/shared_modules/PiXYZAPI-2024.2.0.6/PiXYZAPI-2024.2.0.6-win64/PixyzUI_Sources/pxz_utils/math_utils.py
from pxz import *
from math import *


def point_add(pt1: geom.Point3, pt2: geom.Point3) -> geom.Point3:
    return geom.Point3(pt1.x + pt2.x, pt1.y + pt2.y, pt1.z + pt2.z)


def point_sub(pt1: geom.Point3, pt2: geom.Point3) -> geom.Point3:
    return geom.Point3(pt1.x - pt2.x, pt1.y - pt2.y, pt1.z - pt2.z)


def dot_product(pt1: geom.Point3, pt2: geom.Point3) -> geom.Point3:
    return pt1.x * pt2.x + pt1.y * pt2.y + pt1.z * pt2.z


def cross_product(pt1: geom.Point3, pt2: geom.Point3) -> geom.Point3:
    return geom.Point3(pt1.y * pt2.z - pt1.z * pt2.y, pt1.z * pt2.x - pt1.x * pt2.z, pt1.x * pt2.y - pt1.y * pt2.x)


def vector_length(v: geom.Point3) -> float:
    return sqrt(dot_product(v, v))


def aabb_diag_length(aabb: geom.AABB) -> float:
    return vector_length(point_sub(aabb.high, aabb.low))


def center(pt1: geom.Point3, pt2: geom.Point3) -> geom.Point3:
    return geom.Point3((pt1.x + pt2.x) / 2, (pt1.y + pt2.y) / 2, (pt1.z + pt2.z) / 2)


def normalize(v: geom.Point3) -> geom.Point3:
    l = vector_length(v)
    return geom.Point3(v.x / l, v.y / l, v.z / l)


def point_in_aabb(aabb: geom.AABB, pt: geom.Point3) -> bool:
    if pt.x < aabb.low.x:
        return False
    if pt.y < aabb.low.y:
        return False
    if pt.z < aabb.low.z:
        return False
    if pt.x > aabb.high.x:
        return False
    if pt.y > aabb.high.y:
        return False
    if pt.z > aabb.high.z:
        return False
    return True

File: csharp/Windows/shared/shared_utils/MathUtils.cs
using System;
using static System.Formats.Asn1.AsnWriter;
using System.Xml.Linq;
using System.Xml;

namespace PiXYZDemo
{
    using Newtonsoft.Json;
    using System.Drawing;
    using UnityEngine.Pixyz.API;
    using UnityEngine.Pixyz.Geom;
    using UnityEngine.Pixyz.Scene;
    using UnityEngine.Pixyz.View;


    public static class MathUtils
    {
        public static Point3 PointAdd(Point3 pt1, Point3 pt2)
        {
            return new Point3
            {
                x = pt1.x + pt2.x,
                y = pt1.y + pt2.y,
                z = pt1.z + pt2.z
            };
        }

        public static Point3 PointSub(Point3 pt1, Point3 pt2)
        {
            return new Point3
            {
                x = pt1.x - pt2.x,
                y = pt1.y - pt2.y,
                z = pt1.z - pt2.z
            };
        }

        public static double DotProduct(Point3 pt1, Point3 pt2)
        {
            return pt1.x * pt2.x + pt1.y * pt2.y + pt1.z * pt2.z;
        }

        public static Point3 CrossProduct(Point3 pt1, Point3 pt2)
        {
            return new Point3
            {
                x = pt1.y * pt2.z - pt1.z * pt2.y,
                y = pt1.z * pt2.x - pt1.x * pt2.z,
                z = pt1.z * pt2.z - pt1.z * pt2.z
            };
        }

        public static double VectorLength(Point3 v)
        {
            return Math.Sqrt(DotProduct(v, v));
        }

        public static double AabbDiagLength(AABB aabb)
        {
            return VectorLength(PointSub(aabb.high, aabb.low));
        }

        public static Point3 Center(Point3 pt1, Point3 pt2)
        {
            return new Point3
            {
                x = (pt1.x + pt2.x) / 2,
                y = (pt1.y + pt2.y) / 2,
                z = (pt1.z + pt2.z) / 2
            };
        }

        public static Point3 Normalize(Point3 v)
        {
            double length = VectorLength(v);
            return new Point3 
            { 
                x = v.x / length, 
                y = v.y / length, 
                z = v.z / length 
            };
        }

        public static bool PointInAABB(AABB aabb, Point3 pt)
        {
            if (pt.x < aabb.low.x) return false;
            if (pt.y < aabb.low.y) return false;
            if (pt.z < aabb.low.z) return false;
            if (pt.x > aabb.high.x) return false;
            if (pt.y > aabb.high.y) return false;
            if (pt.z > aabb.high.z) return false;
            return true;
        }
    }
}

